Add support for multiple contiguous-block reads, without the extra overhead of additional `data-copying`

nihalpasham opened the pull request:
Added a method named `read_multi` - reads multiple contiguous blocks of a file in a single read operation, without the extra overhead of additional `data-copying` of the existing `read` method. 

It has the same effect as `[Controller<D, T>::read]` but reduces `read time` by more than 50%, especially in the case of large files (i.e. > 1Mb)

eldruin reviewed (commented):
This is an interesting enhancement. However, it requires the `slice_as_chunks` feature so it requires Rust nightly.
It is a big performance boost, though, so maybe it is worth putting behind a `nightly` feature.

nihalpasham commented:
> This is an interesting enhancement. However, it requires the `slice_as_chunks` feature so it requires Rust nightly. It is a big performance boost, though, so maybe it is worth putting behind a `nightly` feature.

Ok. How do I conditionally add unstable features to a stable build? Is there a standard way to do this.

eldruin commented:
That is not possible, nor should it. However, we can define a new feature like `unstable` (sounds better than `nightly`) or a similar name, that enables features only available on Rust nightly and thus can only be enabled when using it or it will fail to compile.
If/when the feature makes it to stable, we can make it just a no-op.
You can do that with:
```rust
#![cfg_attr(feature = "unstable", feature(slice_as_chunks))]
```
This feature would also need to be documented both in the docs and in the README.

nihalpasham commented:
Ok, I've added the following 

- a new feature named `unstable`  (along with the `cfg_attr` attribute).
- README documentation

It now compiles on stable and nightly (with the unstable feature-flag set). Although, I'm a little unclear as to why this will be a no-op when the `feature` makes it to stable.  The `unstable` feature-flag that decorates `read_multi` would still need to be removed before we use it in `stable-rust` -right? 

PS: I noticed `rust-analyzer` has issues when setting 

```rust
#![cfg_attr(feature = "unstable", feature(slice_as_chunks))]
```
i.e. go to definitions, references within `read_multi` doesn't work. I am assuming this is a `rust-analyzer` bug.

eldruin commented:
Looks good. The CI here needs to be updated to test this feature on rust nightly, though.
Would you be able to do that as well?
I mean if/when this feature is available on rust stable we can make a new non-breaking release where we make the `unstable` feature just do nothing and we make `read_multi` simply available to everybody.

nihalpasham commented:
Fixed the formatting error. Added an additional build step (named `Nightly Build`) to the existing `build-workflow` I hope I got this right (never done this before).

eldruin reviewed (approved):
Looks good to me, thanks!
@thejpster do you have an opinion on this?

thejpster reviewed (commented):
I like the idea of speeding up reads in principle, but I've left a couple of questions, if that's OK.

thejpster commented on src/lib.rs:
```diff
@@ -626,6 +626,105 @@ where
+            };
+
+            self.block_device
+                .read(Block::from_array_slice(blocks), block_idx, "read")
```
The string argument I *think* should be the function name (but it's been a long while since I wrote this so I might be wrong)

nihalpasham replied:
Changed from ~~read~~ to `read_multi`

nihalpasham commented on src/blockdevice.rs:
```diff
@@ -64,6 +64,19 @@ impl Block {
+    /// Converts a [[u8; 512]] slice into a slice of `Block`s,
+    /// assuming that there's no remainder.
+    ///
+    /// # Safety
```
I moved the `Safety` comment into the function.

nihalpasham commented on src/lib.rs:
```diff
@@ -626,6 +626,105 @@ where
+                },
+            },
+        };
+        while (next_cluster.0 - cluster.0) == 1 {
```
I'm not well-versed with the FAT standard, so I may be wrong here but I believe contiguous clusters are (always) in ascending order i.e. if we were to come across a `next cluster` that precedes the `current cluster`, its considered to be a fragmented cluster (and we can simply return 1). 

But in any case, block devices that support `multiple block reads` require a `starting block` and a `block-count`. The implementation can get tricky if we want to account for `next clusters` preceding the current one. So, to avoid additional complexity, I simply assume they are fragmented.

thejpster replied:
Right but if cluster.0 is 4 and next_cluster.0 is 3, this subtraction will overflow before we check == 1. Right?

nihalpasham replied:
Ah, I understand now (was overthinking the question, got carried away with the FAT stuff). Yes, that would result in an overflow. I've updated it to use `wrapping_sub`

nihalpasham commented on src/lib.rs:
```diff
@@ -626,6 +626,105 @@ where
+    /// without the extra overhead of additional `data-copying`.
+    ///
+    /// NOTE: The buffer argument must be a multiple of 512 bytes. This impl assumes the underlying
+    /// emmc driver (and consequently the EMMC device) features support for multi-block reads.
```
- Changed ~~emmc~~ to `block device` and ~~512 bytes~~ to `block size`. 
- Added a note for what happens when the buffer isn't a multiple of block size.
- For why `read_multi` should be a separate method, `read_multi` depends on `hardware support` for multi-block reads. I guess it would make sense to reflect that in the API and avoid checks to determine support for multi-block read (i.e. hardware dependent code).

Note: by hardware support I mean the actual block device (like an sd-card) not so much the host controller.

nihalpasham commented:
Kind of forgot about this. Is this good to go?

thejpster commented:
Can we emulate 'read_multiple' on the block device trait, thus avoiding the need for a feature flag? You could have a default method on the trait which calls the single-block-read function in a loop and memcpy's the data into the given buffer.

nihalpasham commented:
Assuming I understood your question correctly, `emulating` read_multi by looping over multiple `single_block_read` calls technically works but then we don't get the performance 
-  as hardware (i.e. block-device controller) will just end up issuing multiple `single-read` calls. 
- and we still have to deal with additional data-copying. Here's an [example](https://github.com/nihalpasham/rustBoot/blob/c3304a61cb7bc7bf23f7465e58a909b4464cf299/boards/bootloaders/rpi4/src/fs/emmcfat.rs#L101), that avoids additional data-copies.

The extra `feature-flag` is only required to convert a `[[u8; 512]] slice` into a slice of `Block(s)`. I used a nightly only [feature ](https://github.com/rust-lang/rust/issues/74985) here but its possible to implement this ourselves.

thejpster commented:
But you would get the performance if the BlockDevice implementation replaced the default read_multiple implementation with a high-performance version. And a user of the BlockDevice trait (i.e. the FAT filesystem code) wouldn't have to know or care about that.

nihalpasham commented:
So, 

- yes, a block device implementation can add support for reading `multiple-blocks`, if the underlying hardware supports it and it is usually decoupled from the filesystem.
- but for it to be performant i.e. make use of the multiple block-read functionality, its going to need a higher layer (such as the filesystem) to tell it which set of contiguous blocks are to be read. 
- If this info is not provided, then a block device implementation usually falls back to single block reads (much like the current implementation of `[Controller<D, T>::read]`)

thejpster commented:
So my thinking is:

1. The file system always asks for contiguous blocks where it can.
2. The block device provides an accelerated multi-block-read function where possible.
3. The _block device trait_ provides a default implmentation of the multi-block-read function which just calls the single-block-read function in a loop.

The same applies to writes.

thejpster commented:
I'm going to close this, because we didn't get it out of WIP and merged before a whole bunch of stuff happened. Sorry, that's my fault. Feel free to rebase on current head and open a new one.

thejpster commented:
I do want to revisit this though, because I think it is useful to be able to pull larger blocks from the transport in one go.
